Create a FreeBSD package

Crystal looks very interesting. Can we create a package for FreeBSD?

@ysbaddaden I get this error when building v0.23.1 using `crystal-0.19.4-2-freebsd10.1-x86_64.tar.gz` on FreeBSD 11.1-RELEASE.

```
# gmake
Using /usr/local/bin/llvm-config39 [version=3.9.1]         
CRYSTAL_CONFIG_PATH=`pwd`/src ./bin/crystal build  -o .build/crystal src/compiler/crystal.cr -D without_openssl -D without_zlib                                                                                                               
Error in ./src/compiler/crystal.cr:4: while requiring "./crystal/**"                                                   

require "./crystal/**"                                     
^                                                          

in ./src/compiler/crystal/command.cr:11: while requiring "json"                                                        

require "json"                                             
^                                                          

in ./src/json.cr:70: while requiring "./json/*"            

require "./json/*"                                         
^                                                          

in ./src/json/to_json.cr:96: JSON::Builder is not a module, it's a class                                               

module JSON::Builder                                       
       ^                                                   

gmake: *** [Makefile:117: .build/crystal] Error 1          
```

--- a/spec/compiler/crystal/tools/context_spec.cr
+++ b/spec/compiler/crystal/tools/context_spec.cr
@@ -408,14 +408,14 @@ describe "context" do
     ), "self", "a"
   end
 
-  it "can't get a context from never called method" do
+  it "can't get context from uncalled method" do
     run_context_tool %(
     def foo(value)
       ‸
     end
     ) do |result|
       result.status.should eq("failed")
-      result.message.should match(/never called method/)
+      result.message.should match(/never called/)
     end
   end
 end

--- a/spec/compiler/crystal/tools/context_spec.cr
+++ b/spec/compiler/crystal/tools/context_spec.cr
@@ -407,4 +407,15 @@ describe "context" do
     Foo.foo 100
     ), "self", "a"
   end
+
+  it "can't get a context from never called method" do
+    run_context_tool %(
+    def foo(value)
+      ‸
+    end
+    ) do |result|
+      result.status.should eq("failed")
+      result.message.should match(/never called method/)
+    end
+  end
 end

--- a/src/compiler/crystal/tools/context.cr
+++ b/src/compiler/crystal/tools/context.cr
@@ -165,7 +165,7 @@ module Crystal
 
       if @contexts.empty?
         if @found_untyped_def
-          return ContextResult.new("failed", "no context information found (never called method has no context anytime)")
+          return ContextResult.new("failed", "no context information found (methods which are never called don't have a context)")
         else
           return ContextResult.new("failed", "no context information found")
         end

--- a/src/compiler/crystal/tools/context.cr
+++ b/src/compiler/crystal/tools/context.cr
@@ -82,7 +82,9 @@ module Crystal
         defs.each do |typed_def|
           typed_def.accept(self)
           next unless @context_visitor.def_with_yield.not_nil!.location == typed_def.location
-          typed_def.accept(@context_visitor)
+          @context_visitor.inside_typed_def do
+            typed_def.accept(@context_visitor)
+          end
         end
       end
       true
@@ -109,6 +111,13 @@ module Crystal
       @contexts = Array(HashStringType).new
       @context = HashStringType.new
       @def_with_yield = nil
+      @inside_typed_def = false
+      @found_untyped_def = false
+    end
+
+    def inside_typed_def
+      @inside_typed_def = true
+      yield.tap { @inside_typed_def = false }
     end
 
     def process_typed_def(typed_def)
@@ -130,7 +139,7 @@ module Crystal
           add_context ivar.name, ivar.type
         end
       end
-      typed_def.accept(self)
+      inside_typed_def { typed_def.accept(self) }
 
       @contexts << @context unless @context.empty?
     end
@@ -155,7 +164,11 @@ module Crystal
       end
 
       if @contexts.empty?
-        return ContextResult.new("failed", "no context information found")
+        if @found_untyped_def
+          return ContextResult.new("failed", "no context information found (never called method has no context anytime)")
+        else
+          return ContextResult.new("failed", "no context information found")
+        end
       else
         res = ContextResult.new("ok", "#{@contexts.size} possible context#{@contexts.size > 1 ? "s" : ""} found")
         res.contexts = @contexts
@@ -168,6 +181,11 @@ module Crystal
 
       if @def_with_yield.nil? && !node.yields.nil?
         @def_with_yield = node
+        return false
+      end
+
+      unless @inside_typed_def
+        @found_untyped_def = true
         return false
       end
 

